File: src/free_proxy/infrastructure/network/upstream.py
from __future__ import annotations

import os
import urllib.parse
from dataclasses import dataclass

from free_proxy.config import Settings


@dataclass(frozen=True, slots=True)
class UpstreamProxy:
    kind: str
    host: str
    port: int
    username: str | None = None
    password: str | None = None

    @property
    def url(self) -> str:
        auth = ""
        if self.username is not None:
            username = urllib.parse.quote(self.username, safe="")
            password = urllib.parse.quote(self.password or "", safe="")
            auth = f"{username}:{password}@"
        scheme = "socks5" if self.kind == "socks" else "http"
        host = f"[{self.host}]" if ":" in self.host else self.host
        return f"{scheme}://{auth}{host}:{self.port}"


def get_upstream_proxy(settings: Settings) -> UpstreamProxy | None:
    candidates = [
        settings.upstream_proxy_url,
        os.environ.get("http_proxy"),
        os.environ.get("HTTP_PROXY"),
        os.environ.get("https_proxy"),
        os.environ.get("HTTPS_PROXY"),
    ]
    for value in candidates:
        if value:
            parsed = parse_upstream_proxy(value)
            if parsed is not None:
                return parsed
    return None


def parse_upstream_proxy(
    value: str,
    *,
    forced_kind: str | None = None,
) -> UpstreamProxy | None:
    normalized = value.strip()
    if not normalized:
        return None
    if "://" not in normalized:
        normalized = f"{forced_kind or 'http'}://{normalized}"
    parsed = urllib.parse.urlsplit(normalized)
    if not parsed.hostname:
        return None
    kind = forced_kind or ("socks" if parsed.scheme.lower().startswith("socks") else "http")
    # The legacy OpenVPN variables used a SOCKS endpoint on 10808 when no port
    # was specified; explicit URLs retain protocol-specific conventional ports.
    default_port = 1080 if kind == "socks" else 8080
    username = urllib.parse.unquote(parsed.username) if parsed.username is not None else None
    password = urllib.parse.unquote(parsed.password or "") if username is not None else None
    return UpstreamProxy(
        kind=kind,
        host=parsed.hostname,
        port=parsed.port or default_port,
        username=username,
        password=password,
    )

File: internal/netx/upstream.go
package netx

import (
	"fmt"
	"net/url"
	"strconv"
	"strings"
)

// UpstreamProxy describes an optional upstream HTTP/SOCKS proxy used when
// fetching provider data and when dialing OpenVPN over TCP.
type UpstreamProxy struct {
	Kind     string // "http" or "socks"
	Host     string
	Port     int
	Username string
	Password string
	HasAuth  bool
}

// URL renders the proxy as a URL string (with credentials when present).
func (u UpstreamProxy) URL() string {
	auth := ""
	if u.HasAuth {
		auth = url.QueryEscape(u.Username) + ":" + url.QueryEscape(u.Password) + "@"
	}
	scheme := "http"
	if u.Kind == "socks" {
		scheme = "socks5"
	}
	host := u.Host
	if strings.Contains(host, ":") {
		host = "[" + host + "]"
	}
	return fmt.Sprintf("%s://%s%s:%d", scheme, auth, host, u.Port)
}

// GetUpstreamProxy resolves the configured upstream proxy from the config URL
// or standard proxy environment variables (checked by caller-provided list).
func GetUpstreamProxy(configuredURL string, envValues ...string) *UpstreamProxy {
	candidates := append([]string{configuredURL}, envValues...)
	for _, v := range candidates {
		if p := ParseUpstreamProxy(v, ""); p != nil {
			return p
		}
	}
	return nil
}

// ParseUpstreamProxy parses a proxy URL. forcedKind ("http"/"socks") overrides
// scheme detection when set.
func ParseUpstreamProxy(value, forcedKind string) *UpstreamProxy {
	normalized := strings.TrimSpace(value)
	if normalized == "" {
		return nil
	}
	if !strings.Contains(normalized, "://") {
		k := forcedKind
		if k == "" {
			k = "http"
		}
		normalized = k + "://" + normalized
	}
	parsed, err := url.Parse(normalized)
	if err != nil || parsed.Hostname() == "" {
		return nil
	}
	kind := forcedKind
	if kind == "" {
		if strings.HasPrefix(strings.ToLower(parsed.Scheme), "socks") {
			kind = "socks"
		} else {
			kind = "http"
		}
	}
	defaultPort := 8080
	if kind == "socks" {
		defaultPort = 1080
	}
	port := defaultPort
	if p := parsed.Port(); p != "" {
		if n, err := strconv.Atoi(p); err == nil {
			port = n
		}
	}
	out := &UpstreamProxy{Kind: kind, Host: parsed.Hostname(), Port: port}
	if parsed.User != nil {
		out.Username = parsed.User.Username()
		if pw, ok := parsed.User.Password(); ok {
			out.Password = pw
		}
		out.HasAuth = true
	}
	return out
}
